### ISSUE TITLE
PhBaseWorkChain: not recover from ran out of walltime

### ISSUE THREAD

**Issue by unkcpz** [aiidateam/aiida-quantumespresso#753]
My `PhBaseWorkChain` got exit when reaching the walltime. But there is already the walltime handler there in `PhBaseWorkChain` to handle the walltime issue. 

The process report is:

```
2021-11-04 11:58:36 [4706 | REPORT]: [14787|PhBaseWorkChain|run_process]: launching PhCalculation<14790> iteration #1
2021-11-04 13:04:59 [5070 | REPORT]: [14787|PhBaseWorkChain|report_error_handled]: PhCalculation<14790> failed with exit status 120: The job ran out of walltime.
2021-11-04 13:04:59 [5071 | REPORT]: [14787|PhBaseWorkChain|report_error_handled]: Action taken: unrecoverable error, aborting...
2021-11-04 13:04:59 [5072 | REPORT]: [14787|PhBaseWorkChain|inspect_process]: PhCalculation<14790> failed but a handler detected an unrecoverable problem, aborting
2021-11-04 13:04:59 [5073 | REPORT]: [14787|PhBaseWorkChain|on_terminated]: remote folders will not be cleaned
```

versions:
`aiida-core==1.6.5`
`aiida-quantumespresso` current develop branch at https://github.com/aiidateam/aiida-quantumespresso/commit/0874d956c94bba6f2853f8a4984121330b5c77f8

Please tell me how to give you more information to anchor this issue.

**Comment by unkcpz** [comment]
By search `exit status` of `120` in `aiida-core`. I find that it is in `CalcJob` for walltime detect in scheduler:

```
        spec.exit_code(120, 'ERROR_SCHEDULER_OUT_OF_WALLTIME',
            message='The job ran out of walltime.')
```

**Comment by sphuber** [comment]
Yes, this was a newly added feature to `aiida-core`. Scheduler plugins can now add parsing to the scheduler stdout and stderr and return this exit code if they detect an OOW.  If the job was cleanly stopped, we can simply restart the calculation but that should be checked first. The `PhParser` should be updated to check for this exit code and then do additional parsing. If there was a clean stop, it should probably change the exit code to a `400` otherwise leave it at `120`.

**Comment by unkcpz** [comment]
`PhParser` already try to check the `ERROR_OUT_OF_WALLTIME`. I think the problem is that the parser detect `Maximum CPU time exceeded` in output for this exit status while this line is not in the output text. 
So I am curious that with old aiida-core version (before scheduler parser of CalcJob implemented) also can not parse OOW error for ph calculation? Or am I missing something?

If I am going to fix this issue, I guess the update to the `PhParser` should be like this?

```python
if self.node.exit_status == 120:
    return self.exit_codes.ERROR_OUT_OF_WALLTIME
```

**Comment by sphuber** [comment]
There are 2 ways that an OOW can manifest:

1) The job runs out of walltime and so the _scheduler_ kills the job (in this scenario QE does not have the time to shut down cleanly) and there will be an OOW message in the `_scheduler-stderr.txt` file.
2) The input parameters of the `ph.x` calculation specify a `max_wallclock_seconds` (which we do in the `PhBaseWorkChain`, and set it to a fraction of the total walltime requested for the job) and if this is hit, QE will try to shutdown cleanly. In this case, there will be an OOW message at the end of the _stdout_ file. 

In the first case, the scheduler parser will detect the problem and it will set `120` on the node. In the second case, we should indeed set `400`. The difference between `120` and `400` then is that in the latter the code got to shutdown cleanly and we should be able to restart from it. For `120` the restart files are most likely incomplete and unusable, and we have to restart from scratch with a larger walltime.

So, what I think we _should_ do in the `PhParser` is, as you suggest, indeed check for `self.node.exit_status == 120` and in that case, we try to parse as much information as possible, but then in the end simply return the same exit code. (I think just returning `None` should work, as the exit code is already set on the node, but maybe it is better to be explicit and return `return self.node.exit_code`. Here we would have to check if there is a `exit_code` convenience property that reconstructs a `ExitCode` instance from the `exit_status` and `exit_message` attributes that are set on the node.)

Let me know if it is not clear, then maybe we meet quickly in person to discuss it.

**Comment by unkcpz** [comment]
@sphuber thanks! I think it is clear to me your idea of how `PhParser` should be updated, I can have a look and try to fix it.

However, I check the `_aiidasubmit.sh` and `aiida.in` of this calcjob, the wall time in `aiida.in` is within the time set in `_aiidasubmit.sh`. It is supposed to clearly stopped rather than killed by _scheduler_, correct? Or is 90 seconds in this case is still not enough for the calculation to shutdown clearly?

```
jyu@imxgesrv1:/exports/commonscratch/jyu/aiida_run/8b/fb/a7fa-857d-483a-bd3a-a363259b2874$ cat aiida.in 
&INPUTPH
  epsil = .false.
  fildyn = 'DYN_MAT/dynamical-matrix-'
  ldisp = .false.
  max_seconds =   1.7100000000d+03
  outdir = './out/'
  prefix = 'aiida'
  recover = .true.
  tr2_ph =   1.0000000000d-16
  verbosity = 'high'
/
      0.0000000000       0.6123724357       0.0000000000
```

```
jyu@imxgesrv1:/exports/commonscratch/jyu/aiida_run/8b/fb/a7fa-857d-483a-bd3a-a363259b2874$ cat _aiidasubmit.sh 
#!/bin/bash
#SBATCH --no-requeue
#SBATCH --job-name="aiida-14790"
#SBATCH --get-user-env
#SBATCH --output=_scheduler-stdout.txt
#SBATCH --error=_scheduler-stderr.txt
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=32
#SBATCH --time=00:30:00


#SBATCH --cpus-per-task=2

export OMP_NUM_THREADS=${SLURM_CPUS_PER_TASK}


module load intel/19.1.3.304
module load intel-mkl/2020.4.304
module load intel-mpi/2019.9.304
module load quantum-espresso/6.5.0-mpi


'srun' '-n' '32' '/home/software/apps/codes/intel/19.1.3.304/quantum-espresso/6.5.0_MPI_MKL/bin/ph.x' '-in' 'aiida.in'  > 'aiida.out' 
```

**Comment by sphuber** [comment]
> Or is 90 seconds in this case is still not enough for the calculation to shutdown clearly?

When the `max_seconds` is passed, `ph.x` will stop the calculation and start writing the data from memory to disk. If this takes more time than is remaining for the job, the scheduler can _still_ kill it. So one possible solution for the `PhBaseWorkChain` is to try and increase the `max_seconds` set in the input parameters (so not the `max_wallclock_seconds` in the `metadata.options.resources` which I think we shouldn't just change like that) and run again. Maybe with more time, it can actually write the files to disk before the walltime is used up.

**Comment by unkcpz** [comment]
> So one possible solution for the PhBaseWorkChain is to try and **increase** the max_seconds set in the input parameters (so not the max_wallclock_seconds in the metadata.options.resources which I think we shouldn't just change like that) and run again.

I guess you mean to say **decrease** the max_seconds in the input parameters? The max_seconds hit and start to stop the calculation before the scheduler try to kill the job.

**Comment by sphuber** [comment]
Yes, of course, my bad. I meant to say decrease `max_seconds`.
